Adds details on how to customize external access (#897)
redpanda-data/documentation: docs/platform/quickstart/kubernetes-qs-dev.mdx

| 31644 | 9644 | Admin API |

You can modify these ports in `values.yaml`. If you do so, then make sure to use the same ports in the steps below.

### Modify inbound security

Cloud-based Kubernetes clusters will have a security group tied to each node in the cluster. For instance, in EKS a security group is applied to each EC2 node.

You must add inbound rules allowing traffic to reach each port above on all nodes that are part of the cluster to enable external access. How these rules are added depends on your Kubernetes environment and cloud provider.

### Update external access

By default the cluster is available externally at `redpanda-<replica index>.<redpanda.external.domain>`. The default values would make this address `redpanda-0.local`. But no external DNS provider is updated to make this address resolvable by default. There are a few ways to resolve this:

1. Use IP address resolution
2. Make `redpanda-0.local` resolvable to the external address of the related Kubernetes node
3. Customize the subdomain and/or domain to point to something resolvable via DNS

Since options 2 and 3 require an external DNS provider, this guide focuses on option 1.

> Each Kubernetes node has an external IP address, and they can be found with the following command:

Reviewer: In our customer success workshop, we had to specify the path to the values.yaml file --needed to pull redpanda out here and replace with `helm upgrade redpanda redpanda/redpanda -n redpanda -f values.yaml`

Reviewer: Suggestion was also made to add a Prereqs section here - AWS CLI must be installed. Add any other prereqs customer needs before setting up external access.

Reviewer 2: I have this change locally as I'm also working on this doc. It may be simpler to get this merged, then I can submit my changes.

redpanda-0                    1/1     Running     0          109m   192.168.30.134   ip-192-168-31-253.us-east-2.compute.internal   <none>           <none>
redpanda-1                    1/1     Running     0          109m   192.168.60.112   ip-192-168-41-207.us-east-2.compute.internal   <none>           <none>
redpanda-2                    1/1     Running     0          110m   192.168.68.55    ip-192-168-90-157.us-east-2.compute.internal   <none>           <none>
redpanda-post-install-skkl7   0/1     Completed   0          110m   192.168.73.172   ip-192-168-90-157.us-east-2.compute.internal   <none>           <none>
redpanda-post-upgrade-r85f6   0/1     Completed   0          110m   192.168.33.87    ip-192-168-41-207.us-east-2.compute.internal   <none>           <none>
```

The above example shows `redpanda-0` running on a node with the name `ip-192-168-31-253.us-east-2.compute.internal`, giving us the internal IP address of `192.168.31.253`. Using this combined with the output from the previous command, you get the following associations:

> | Redpanda node | External IP |

Reviewer 2: Is it the Redpanda node or the Redpanda broker that we are referring to here?

From our style guide, I think it's the broker: https://vectorizedio.atlassian.net/wiki/spaces/DOC/pages/92635182/Redpanda+Documentation+Style+Guide#Word-list

Author: Looks like that doc doesn't actually say to use broker over node... node is listed below. And if we go with broker over node then there is a ton more docs (and code) that would also need to be changed.

Reviewer 2: I understand. But, especially in the Kubernetes world, mixing these terms is likely to confuse users, so it would be good in the future to standardize them and be explicit about how we use them.

Kubernetes nodes vs Redpanda nodes vs Redpanda brokers vs Redpanda Pods

@joejulian 's doc for example
https://github.com/redpanda-data/helm-charts/edit/main/examples/with-external-via-nodeport/README.md

> These instructions assume you have 3 replicas, one pod (and by extension Redpanda broker) per node

Author: > Kubernetes nodes vs Redpanda nodes vs Redpanda brokers vs Redpanda Pods

Here is my understanding:

- Kubernetes node: this is an easy definition (more details [here](https://kubernetes.io/docs/concepts/architecture/nodes/)
- Redpanda node/broker: I can't tell the difference between these two things so I have to put them on the same line. I almost feel like the term "Redpanda node" shouldn't exist in the same world with the term "Redpanda broker". I guess I would choose broker over node in spite of the term node being used throughout the codebase, but then we'd have to be clear that a Redpanda broker has many additional capabilities that a Kafka broker doesn't have (and this should be easy to do). I've seen the line about "...multiple brokers can run on a node..." but it feels like the author of that was confusing Kubernetes nodes with Redpanda nodes.
- Redpanda pods: This doesn't exist. Kubernetes pods sometimes run Kubernetes nodes with Redpanda nodes/brokers, but pod is definitely a Kubernetes thing

As for standardizing on either node or broker, I totally agree and I know others do to (hi @joejulian ). But resolving this has proven difficult so far.

Reviewer 2: Thanks @vuldin ! I'd prefer us to standardize the use of Redpanda broker to avoid the confusion of Redpanda nodes vs Kubernetes nodes, especially when the qualifier in omitted.

|---|---|
| redpanda-0 | 203.0.113.3 |
| redpanda-1 | 203.0.113.5 |
| redpanda-2 | 203.0.113.7 |

:::note
The Kubernetes node order and Redpanda node order were the same in the above example, but this will not always be the case.
:::

Now update `values.yaml` with the external IPs, making sure to match the Redpanda node order:

```yaml
redpanda:
  external:
    addressType: ip

>     addresses: "203.0.113.3 203.0.113.5 203.0.113.7"

Reviewer 2: So is the order here important and why?

> making sure to match the Redpanda node order

Do the IP addresses have to go in the same order as the ordinals in the name of each Pod in the StatefulSet? And, if so, how come?

Author: The order is important. These IPs are parsed, and then each are assigned to Redpanda nodes starting with 0 and on to the last. This is how we know that `redpanda-0` is associated with `203.0.113.3` (for example).

Author: When using IP addresses I guess it doesn't matter that much. The association would be wrong, but as long as all nodes are included when sent to clients then maybe there wouldn't be an issue.

But `addresses` could be subdomains when `addressType` is set to `subdomain`. In that case it would be more obvious the association is incorrect. This mismatch could result in more redirects to the proper node. I'm not sure if there are any other issues that could pop up, definitely would make any investigations more difficult. For instance, if you need to check the leader logs and you run `rpk cluster info`, then it would report the address of the leader node. But then looking at the logs and using `kubectl` would give you different info.

```text
statefulset rolling update complete 3 pods at revision redpanda-8654f645b4...
```

> ## External access with AWS EKS

Reviewer 2: I think we should be consistent with the rest of the doc and make the heading generic.

We can use tabs for specific platforms.

The following steps assume you followed the EKS route in the steps above.

### Modify security group

> EKS creates a security group applied to each EC2 node in the cluster. Additional inbound rules are needed on this security group to allow external access to the Redpanda nodes through the deployed NodePort service.

Reviewer 2: Could we explain why we expose only the Kafka and admin API ports in this step?

Author: I'll change this to instruct the user to add rules for each of the NodePort service's handled ports.

1. Open AWS Console and navigate to EC2 > Instances
2. Select one of the EC2 nodes associated with your cluster (check nodes in the EKS view if needed)
3. Select the Security tab

> 4. Select "Security group"

Reviewer 2: I don't think we should document a third-party tool like this. It's difficult to maintain if Amazon decide to change things.

It'd be better to tell users what to do and provide a link to the AWS docs.

Something like: 'Add the following inbound rules to your security group.'

Reviewer: Agree--we do not ever document other products. Had this issue with cloud features and AWS/GCP/Azure in the past at other companies. Best practice is to link to the step in that company's product doc. 
I have heard arguments against this, but the bottom line is the RP docs does not own that content--including it here would mean that we are responsible for any user bugs wrt that AWS doc. What if AWS adds/deletes/updates that content, or changes the link? RP writers would be responsible for tracking and making both RP features updates and the respective platform doc updates. This is common practice is doc orgs.

5. Select "Edit inbound rules"
6. Select "Add rule" then enter the following details: Type:Custom TCP, Port:31092, Source:Anywhere-IPv4
7. Select "Add rule" then enter the following details: Type:Custom TCP, Port:31644, Source:Anywhere-IPv4
7. Select "Save rules"

:::note

> Step 6 enables external access to the Kafka and admin API ports, but there are other ports made available by the NodePort service. Run the following command for details on the NodePort service:

Reviewer 2: I don't think this warrants an info callout. It's a teachable moment, so it should be a step in the overall procedure above. Something like:

1. Find out which ports are made available by the NodePort service.

   ```bash
   kubectl get service redpanda-external -n redpanda
   ```
   Example output:
   ```
   ...output of kubectl...
   ```

3. Make a note of the Kafka API port and the admin API port. You expose these ports in later steps.

NAME                TYPE       CLUSTER-IP      EXTERNAL-IP   PORT(S)                                                       AGE
redpanda-external   NodePort   10.100.15.142   <none>        9644:31644/TCP,9094:31092/TCP,8083:30082/TCP,8080:30081/TCP   22m
```

As shown above, the following ports are what is opened by the NodePort service by default:

| Node port | Pod port | Purpose |
|---|---|---|
| 30081 | 8080 | Schema registry |
| 30082 | 8083 | HTTP proxy |
| 31092 | 9094 | Kafka API |
| 31644 | 9644 | admin API |

Repeat the steps above to open the additional ports for the HTTP proxy and/or schema registry if desired.
:::

### Update external access via Helm

By default the cluster is available externally via `redpanda-<replica index>.<redpanda.external.domain>`. The default values would make this address `redpanda-0.local`. But no external DNS provider is updated to make this address resolvable by default. There are a few ways to resolve this:

1. Use IP address resolution
2. Make `redpanda-0.local` resolvable to the external address of the related Kubernetes node
3. Customize the subdomain and/or domain to point to something resolvable via DNS

> Since options 2 and 3 require an external DNS provider, we'll focus on option 1 in this guide.

Reviewer 2: ```suggestion
Since options 2 and 3 require an external DNS provider, this guide focuses on option 1.
```

redpanda-0                    1/1     Running     0          109m   192.168.30.134   ip-192-168-31-253.us-east-2.compute.internal   <none>           <none>
redpanda-1                    1/1     Running     0          109m   192.168.60.112   ip-192-168-41-207.us-east-2.compute.internal   <none>           <none>
redpanda-2                    1/1     Running     0          110m   192.168.68.55    ip-192-168-90-157.us-east-2.compute.internal   <none>           <none>
redpanda-post-install-skkl7   0/1     Completed   0          110m   192.168.73.172   ip-192-168-90-157.us-east-2.compute.internal   <none>           <none>
redpanda-post-upgrade-r85f6   0/1     Completed   0          110m   192.168.33.87    ip-192-168-41-207.us-east-2.compute.internal   <none>           <none>
```

The above example shows `redpanda-0` running on a node with the name `ip-192-168-31-253.us-east-2.compute.internal`, giving us the internal IP address of `192.168.31.253`. Using this combined with the output from the previous command, we get the following associations:

| Redpanda node | External IP |
|---|---|

> | redpanda-0 | 3.129.52.203 |

Reviewer 2: are those AWS IPs static? Or would we need additional steps to explain how to do that?

Author: These external IPs are static, as they are associated with the EC2/Kubernetes nodes.

| redpanda-1 | 3.12.164.46 |
| redpanda-2 | 3.144.145.170 |

:::note

> The Kubernetes node order and Redpanda node order were the same in the above example, But this is not always the case!

Reviewer: ```suggestion
The Kubernetes node order and Redpanda node order were the same in the above example, But this is not always the case.
```

:::

Now update `values.yaml` with the external IPs, making sure to match the Redpanda node order. This will require modifying `nodeType` and `nodes` as shown below:

```yaml
redpanda:
  external:
    nodes: "3.129.52.203 3.12.164.46 3.144.145.170"
    nodeType: ip
```

> Once these updates are made, then apply it via the following helm command:

Reviewer: ```suggestion
Once these updates are made, then apply it using the following helm command:
```

```
helm upgrade redpanda redpanda/redpanda -n redpanda
```

### Verify external access

We should now be able to connect to our cluster from an external client, and we'll use a locally-installed `rpk` client (details on how to install are [here](https://docs.redpanda.com/docs/platform/quickstart/rpk-install/)). Set the `REDPANDA_BROKERS` environment variable using the external IP addresses from above:

```
export REDPANDA_BROKERS=3.129.52.203:31092,3.12.164.46:31092,3.144.145.170:31092
```

Get the cluster info:

```

> > rpk cluster info

Reviewer 2: It's best to separate the command from the output to allow users to use the copy/paste button in the docs.
